### PR TITLE
Modify config option for SHA384.

### LINKED ISSUE #4034: Convert negative to positive options
Convert negative options in config.h to positive style, e.g.

MBEDTLS_NO_64BIT_MULTIPLICATION becomes MBEDTLS_HAVE_64BIT_MULTIPLICATION (and uses of this need to be inverted)

### LINKED ISSUE #4472: Change SHA1 define guard to SHA512 in ssl_ciphersuites.c
_Originally posted by @gilles-peskine-arm in https://github.com/ARMmbed/mbedtls/pull/4304#discussion_r625742279_

in library/ssl_ciphersuites.c (https://github.com/ARMmbed/mbedtls/blob/development/library/ssl_ciphersuites.c#L854), there is a mistake in the `#if defined` code guard for SHA1. #4304 for the 3.0 release will change it to SHA384. In 2.16 and 2.2x this define guard should be for SHA512.

### LINKED ISSUE #4499: SHA384 cipher suites are offered when MBEDTLS_SHA512_NO_SHA384 is enabled
When the configuration option `MBEDTLS_SHA512_NO_SHA384` is enabled, the SHA-384 algorithm is disabled, and in particular TLS cipher suites that use SHA-384 cannot be used. However they are still offered by `mbedtls_ssl_list_ciphersuites`, leading to an internal error when selected.

Steps to reproduce:
```
$ scripts/config.py set MBEDTLS_SHA512_NO_SHA384
$ make
$ programs/ssl/ssl_client2 force_ciphersuite=TLS-ECDH-ECDSA-WITH-AES-256-GCM-SHA384

  . Seeding the random number generator... ok
  . Loading the CA root certificate ... ok (0 skipped)
  . Loading the client cert. and key... ok (key type: RSA)
  . Connecting to tcp/localhost/4433... ok
  . Setting up the SSL/TLS structure... ok
  . Performing the SSL/TLS handshake... failed
  ! mbedtls_ssl_handshake returned -0x6c00

Last error was: -0x6C00 - SSL - Internal error (eg, unexpected failure in lower-level module)

```

This only applies to 2.2x. In 2.16, `MBEDTLS_SHA512_NO_SHA384` did not exist yet. In 3.0 this is fixed by #4304.

### PR DESCRIPTION
**Pull request by mstarzyk-mobica**
Instead of using it as MBEDTLS_SHA512_NO_SHA384 which disables
SHA384, now it can be enabled in the config
option MBEDTLS_SHA512_WITH_SHA384.

Resolves #4034

This PR also resolves issues discovered while working on it:
* Fix #4472 (backported in #4503, #4504).
* Fix #4499 (backported in #4503, not applicable to 2.16).

### CODE REVIEW

**Comment by gilles-peskine-arm** [comment]
CI failures as of bc90982: here are the logs from `test_default_cmake_gcc_asan`: 
[test_default_cmake_gcc_asan.txt](https://github.com/ARMmbed/mbedtls/files/6304778/test_default_cmake_gcc_asan.txt)

Many other jobs have failures, I think it's the same test cases but haven't looked thoroughly.

File: tests/scripts/depends-hashes.pl [hunk 1]
```diff
@@ -67,27 +71,29 @@ sub abort {
...
     print "\n******************************************\n";
-    print "* Testing hash option: $hash\n";
+    print "* Testing hash options: ", join(', ', @$hash_config), "\n";
     print "******************************************\n";
-    $ENV{MBEDTLS_TEST_CONFIGURATION} = "-$hash";
 
-    system( "scripts/config.py $hash" )
-        and abort "Failed to $hash\n";
+    for my $hash (@$hash_config) {
+        $ENV{MBEDTLS_TEST_CONFIGURATION} = "-$hash";
```
Comment (gilles-peskine-arm) [inline tests/scripts/depends-hashes.pl:83]: This is setting an environment variable. It shouldn't be in the inner loop. The value of `MBEDTLS_TEST_CONFIGURATION` is only used in test logs, so its exact content doesn't matter. I think `$hash_config[0]` would be a good choice.

File: tests/scripts/depends-hashes.pl [hunk 2]
```diff
@@ -67,27 +71,29 @@ sub abort {
...
     for my $opt (@ssl) {
         system( "scripts/config.py unset $opt" )
             and abort "Failed to disable $opt\n";
     }
 
     system( "CFLAGS='-Werror -Wall -Wextra' make lib" )
-        and abort "Failed to build lib: $hash\n";
-    system( "cd tests && make" ) and abort "Failed to build tests: $hash\n";
-    system( "make test" ) and abort "Failed test suite: $hash\n";
+        and abort "Failed to build lib: @$hash_config\n";
```
Comment (gilles-peskine-arm) [inline tests/scripts/depends-hashes.pl:94]: `@$hash_config` results in something like `unset MBEDTLS_SHA512_Cunset MBEDTLS_SHA384_C`. Better make this `join(', ', @$hash_config)` as above (put this in a local variable).

File: tests/scripts/depends-hashes.pl [hunk 3]
```diff
@@ -50,14 +50,18 @@
...
-
-# list hash options with corresponding actions
-my @hashes = ((map { "unset $_" } @hash_modules));
+# Each element of this array holds list of configuration options that
+# should be tested together. Certain options depend on eachother and
+# separating them would generate invalid configurations.
+my @hash_configs = (
+    ['unset MBEDTLS_MD2_C'],
+    ['unset MBEDTLS_MD4_C'],
+    ['unset MBEDTLS_MD5_C'],
```
Comment (gilles-peskine-arm) [inline tests/scripts/depends-hashes.pl:59]: Missing `RIPEMD160`

Comment (gilles-peskine-arm) [reply tests/scripts/depends-hashes.pl:59]: Oh, that's preexisting! Never mind then. We can improve this later (one day™ we'll get rid of this script anyway in favor of a unified `depends.py`).

**Comment by mstarzyk-mobica** [comment]
fixed conflict with rebase

**Comment by mpg** [comment]
I'm seeing to reviewers here, so I'm removing the label "needs: reviewer". @gilles-peskine-arm @ronald-cron-arm please shout if I misread the situation.

**Comment by ronald-cron-arm** [comment]
I am on it.

**Review by ronald-cron-arm — CHANGES_REQUESTED** [review]
Please find my review comments below. Regarding completeness, MBEDTLS_SHA512_NO_SHA384 not present in the code base anymore. Otherwise when I do:
1) rgrep -n2 MBEDTLS_SHA512_C and search for instances of SHA384 the following ones are suspicious:
--
include/mbedtls/ssl.h:830:#elif defined(MBEDTLS_SHA512_C)                       
include/mbedtls/ssl.h-831-#define MBEDTLS_SSL_PEER_CERT_DIGEST_DFL_TYPE MBEDTLS_MD_SHA384
include/mbedtls/ssl.h-832-#define MBEDTLS_SSL_PEER_CERT_DIGEST_DFL_LEN  48      
--
library/ssl_cookie.c:51:#elif defined(MBEDTLS_SHA512_C)                         
library/ssl_cookie.c-52-#define COOKIE_MD           MBEDTLS_MD_SHA384           
library/ssl_cookie.c-53-#define COOKIE_MD_OUTLEN    48                          
--
library/oid.c:360:#if defined(MBEDTLS_SHA512_C)                                 
library/oid.c-361-    {                                                         
library/oid.c-362-        { ADD_LEN( MBEDTLS_OID_PKCS1_SHA384 ),     "sha384WithRSAEncryption",  "RSA with SHA-384" },
and there are a couple of other ones in oid.c. Please check all MBEDTLS_SHA512_C in oid.c.
--
library/ssl_tls.c:762:#if defined(MBEDTLS_SHA512_C)                             
library/ssl_tls.c-763-        case MBEDTLS_SSL_TLS_PRF_SHA384:                  
library/ssl_tls.c-764-            tls_prf = tls_prf_sha384;                     
library/ssl_tls.c-765-        break;                                            
library/ssl_tls.c:766:#endif /* MBEDTLS_SHA512_C */
--
library/ssl_tls.c:1310:#if defined(MBEDTLS_SHA512_C)                            
library/ssl_tls.c-1311-    if( minor_ver == MBEDTLS_SSL_MINOR_VERSION_3 &&      
library/ssl_tls.c-1312-        hash == MBEDTLS_MD_SHA384 )
and there are a couple of other ones in ssl_tls.c. Please check all MBEDTLS_SHA512_C in ssl_tls.c.
--

2) rgrep -n2 MBEDTLS_SHA256_C and search for instances of SHA224 the following ones are suspicious:

tests/suites/test_suite_shax.data-53-                                           
tests/suites/test_suite_shax.data-54-SHA-224 Test Vector NIST CAVS #1           
tests/suites/test_suite_shax.data:55:depends_on:MBEDTLS_SHA256_C                
tests/suites/test_suite_shax.data-56-sha224:"":"d14a028c2a3a2bc9476102bb288234c415a2b01f828ea62ac5b3e42f"
tests/suites/test_suite_shax.data-57-                                           
tests/suites/test_suite_shax.data-58-SHA-224 Test Vector NIST CAVS #2           
tests/suites/test_suite_shax.data:59:depends_on:MBEDTLS_SHA256_C                
tests/suites/test_suite_shax.data-60-sha224:"ff":"e33f9d75e6ae1369dbabf81b96b4591ae46bba30b591a6b6c62542b5"
tests/suites/test_suite_shax.data-61-                                           
tests/suites/test_suite_shax.data-62-SHA-224 Test Vector NIST CAVS #3           
tests/suites/test_suite_shax.data:63:depends_on:MBEDTLS_SHA256_C                
tests/suites/test_suite_shax.data-64-sha224:"984c":"2fa9df9157d9e027cfbc4c6a9df32e1adc0cbe2328ec2a63c5ae934e"
tests/suites/test_suite_shax.data-65-                                           
tests/suites/test_suite_shax.data-66-SHA-224 Test Vector NIST CAVS #4           
tests/suites/test_suite_shax.data:67:depends_on:MBEDTLS_SHA256_C                
tests/suites/test_suite_shax.data-68-sha224:"50efd0":"b5a9820413c2bf8211fbbf5df1337043b32fa4eafaf61a0c8e9ccede"
tests/suites/test_suite_shax.data-69-                                           
tests/suites/test_suite_shax.data-70-SHA-224 Test Vector NIST CAVS #5           
tests/suites/test_suite_shax.data:71:depends_on:MBEDTLS_SHA256_C                
tests/suites/test_suite_shax.data-72-sha224:"e5e09924":"fd19e74690d291467ce59f077df311638f1c3a46e510d0e49a67062d"
tests/suites/test_suite_shax.data-73-                                           
tests/suites/test_suite_shax.data-74-SHA-224 Test Vector NIST CAVS #6           
tests/suites/test_suite_shax.data:75:depends_on:MBEDTLS_SHA256_C                
tests/suites/test_suite_shax.data-76-sha224:"21ebecb914":"78f4a71c21c694499ce1c7866611b14ace70d905012c356323c7c713"
tests/suites/test_suite_shax.data-77-                                           
tests/suites/test_suite_shax.data-78-SHA-224 Test Vector NIST CAVS #7           
tests/suites/test_suite_shax.data:79:depends_on:MBEDTLS_SHA256_C

File: configs/config-psa-crypto.h [hunk 4]
```diff
@@ -1249,6 +1249,17 @@
  */
 //#define MBEDTLS_SHA256_SMALLER
 
+/**
+ * \def MBEDTLS_SHA384_C
```
[the remark this answers is not in the document]
Comment (gilles-peskine-arm) [reply configs/config-psa-crypto.h:1253]: We (I forget who exactly) discussed it in chat and decided to go with `MBEDTLS_SHA384_C` and `MBEDTLS_SHA224_C` for uniformity with other hashes. The fact that the same `.c` file happens to define both SHA-512 and SHA-384 is an implementation detail. We decided that this was a good reason to deviate from the historical practice of having `MBEDTLS_XXX_C` correspond to `xxx.c`, which already wasn't always followed (TLS, PSA, `ecp_curves`, …).

File: tests/suites/test_suite_x509write.data [hunk 5]
```diff
@@ -3,15 +3,15 @@ depends_on:MBEDTLS_SHA1_C:MBEDTLS_RSA_C:MBEDTLS_PKCS1_V15
 x509_csr_check:"data_files/server1.key":"data_files/server1.req.sha1":MBEDTLS_MD_SHA1:0:0:0:0
```
Comment (gilles-peskine-arm) [inline tests/suites/test_suite_x509write.data:4]: Please update dependencies in `test_suite_x509parse.data` as well.

Note: strictly speaking updating dependencies that don't break the CI is out of scope. However, it's easy to deal with them all, and you've already handled most of them. So I'm requesting to handle them all.

I used the following one-liner to look for problematic test cases:
```
perl -00 -ne '/sha.?(224|384)/i or next; $h=$1; /\ndepends_on:.*_SHA_?$h.*\n.*sha_?$h/i and next; print' tests/suites/test_suite_*.data
```
There may be a few false positive and false negatives, but it's close enough to be useful.

File: tests/suites/test_suite_rsa.data [hunk 6]
```diff
@@ -50,7 +50,7 @@ depends_on:MBEDTLS_SHA1_C:MBEDTLS_PKCS1_V15
 mbedtls_rsa_pkcs1_verify:"55013a489e09b6553262aab59fb041b49437b86d52876f8e5d5e405b77ca0ff6ce8ea2dd75c7b3b411cf4445d56233c5b0ff0e58c49128d81b4fedd295e172d225c451e13defb34b87b7aea6d6f0d20f5c55feb71d2a789fa31f3d9ff47896adc16bec5ce0c9dda3fde190e08ca2451c01ff3091449887695f96dac97ad6a30e":MBEDTLS_RSA_PKCS_V15:MBEDTLS_MD_SHA1:1024:16:"e28a13548525e5f36dccb24ecb7cc332cc689dfd64012604c9c7816d72a16c3f5fcdc0e86e7c03280b1c69b586ce0cd8aec722cc73a5d3b730310bf7dfebdc77ce5d94bbc369dc18a2f7b07bd505ab0f82224aef09fdc1e5063234255e0b3c40a52e9e8ae60898eb88a766bdd788fe9493d8fd86bcdd2884d5c06216c65469e5":16:"10001":"dd82b7be791c454fbbf6f1de47cbe585a687e4e8bbae0b6e2a77f8ca4efd06d71498f9a74b931bd59c377e71daf708a624c51303f377006c676487bad57f7067b09b7bb94a6189119ab8cf7321c321b2dc7df565bfbec833a28b86625fb5fd6a035d4ed79ff0f9aee9fa78935eec65069439ee449d7f5249cdae6fdd6d8c2a63":MBEDTLS_ERR_RSA_VERIFY_FAILED
 
 RSA PKCS1 Verify v1.5 CAVS #11
-depends_on:MBEDTLS_SHA256_C:MBEDTLS_PKCS1_V15
+depends_on:MBEDTLS_MBEDTLS_SHA224_CSHA256_C:MBEDTLS_PKCS1_V15
```
[the remark this answers is not in the document]
Comment (gilles-peskine-arm) [reply tests/suites/test_suite_rsa.data:53]: Memo to one of us: review the test outcome file on the Jenkins CI. It lists test cases that are not executed in any configuration, which lets us identify misspelled dependencies in tests.

**Review by gilles-peskine-arm — CHANGES_REQUESTED** [review]
Overall this looks mostly good. There are a few mistakes in some documentation and dependencies.

File: configs/config-psa-crypto.h (hunk 4 above)
[the remark this answers is not in the document]
Comment (ronald-cron-arm) [reply configs/config-psa-crypto.h:1253]: ok, thanks for the response. Fine by me, resolving this conversation.

File: library/psa_crypto_hash.c [hunk 7]
```diff
@@ -96,15 +96,15 @@ const mbedtls_md_info_t *mbedtls_md_info_from_psa( psa_algorithm_t alg )
...
+#if defined(MBEDTLS_SHA224_C)
         case PSA_ALG_SHA_224:
             return( &mbedtls_sha224_info );
 #endif
 #if defined(MBEDTLS_SHA256_C)
         case PSA_ALG_SHA_256:
             return( &mbedtls_sha256_info );
 #endif
-#if defined(MBEDTLS_SHA512_C) && !defined(MBEDTLS_SHA512_NO_SHA384)
+#if defined(MBEDTLS_SHA384_C)
```
[the remark this answers is not in the document]
Comment (mstarzyk-mobica) [reply library/psa_crypto_hash.c:107]: current ordering is:
224, 256, 384, 512.

why change it to:
224, 384, 256, 512 ??

File: tests/ssl-opt.sh [hunk 8]
```diff
@@ -4302,6 +4304,7 @@ run_test    "Authentication, CA callback: server ECDH p256v1, client optional, p
             -c "! Certificate verification flags"\
             -c "bad server certificate (ECDH curve)" # Expect failure only at ECDH params check
 
+requires_config_enabled MBEDTLS_SHA384_C
```
[the remark this answers is not in the document]
Comment (mstarzyk-mobica) [reply tests/ssl-opt.sh:4307]: I believe this one was failing for some reason, although I cannot recall what it was. I think adding this resolved the issue, but I'll remove it now and verify if it still occurs, it's possible it's not really necessery. Thanks for spotting this.

[the remark this answers is not in the document]
Comment (gilles-peskine-arm) [reply tests/ssl-opt.sh:4307]: There's already a mechanism to skip test cases that require a ciphersuite if that ciphersuite is not enabled. See `maybe_requires_ciphersuite_enabled`. So adding explicit dependencies just because of the ciphersuite should not be required.

File: configs/config-ccm-psk-tls1_2.h [hunk 9]
```diff
@@ -49,6 +49,7 @@
 #define MBEDTLS_MD_C
 #define MBEDTLS_NET_C
 #define MBEDTLS_SHA256_C
+#define MBEDTLS_SHA224_C
```
[the remark this answers is not in the document]
Comment (mstarzyk-mobica) [reply configs/config-ccm-psk-tls1_2.h:52]: I thought that every config which has 256 must have 224. At least that's the condition implemented in the check-config header file.

[the remark this answers is not in the document]
Comment (gilles-peskine-arm) [reply configs/config-ccm-psk-tls1_2.h:52]: Ah, good point. Please add a comment stating that SHA-224 is only enabled because the code base doesn't support enabling SHA-256 without SHA-224, and a future version will remove SHA-224.

[the remark this answers is not in the document]
Comment (mstarzyk-mobica) [reply configs/config-ccm-psk-tls1_2.h:52]: Ok, will do :)

File: configs/config-psa-crypto.h [hunk 10]
```diff
@@ -2624,23 +2635,43 @@
...
  * This module is required for the SSL/TLS 1.2 PRF function.
  */
 #define MBEDTLS_SHA256_C
 
+/**
+ * \def MBEDTLS_SHA224_C
+ *
+ * Enable the SHA-224 cryptographic hash algorithms.
+ *
+ * Module:  library/sha256.c
```
[the remark this answers is not in the document]
Comment (mstarzyk-mobica) [reply configs/config-psa-crypto.h:2649]: As per @gilles-peskine-arm  comment, I'll leave `module` and `caller` in 224 and add it to 384.

[the remark this answers is not in the document]
Comment (ronald-cron-arm) [reply configs/config-psa-crypto.h:2649]: Yes better.

File: library/ssl_cookie.c [hunk 11]
```diff
@@ -44,7 +44,7 @@
  * available. Try SHA-256 first, 512 wastes resources since we need to stay
  * with max 32 bytes of cookie for DTLS 1.0
  */
-#if defined(MBEDTLS_SHA256_C)
+#if defined(MBEDTLS_SHA224_C)
```
[the remark this answers is not in the document]
Comment (mstarzyk-mobica) [reply library/ssl_cookie.c:47]: Since 224 and 256 are bound to each other, we can expect `MBEDTLS_SHA224_C` to be defined. At least until next library version. Are you sure this should be changed back to `#if defined(MBEDTLS_SHA256_C)` ?

File: library/ssl_ciphersuites.c [hunk 12]
```diff
@@ -851,13 +853,13 @@ static const mbedtls_ssl_ciphersuite_t ciphersuite_definitions[] =
       0 },
 #endif /* MBEDTLS_SHA256_C */
 
-#if defined(MBEDTLS_SHA1_C)
+#if defined(MBEDTLS_SHA384_C)
```
[the remark this answers is not in the document]
Comment (mstarzyk-mobica) [reply library/ssl_ciphersuites.c:856]: #4472

File: library/ssl_cookie.c (hunk 11 above)
[the remark this answers is not in the document]
Comment (gilles-peskine-arm) [reply library/ssl_cookie.c:47]: Ok, let's keep 224 for now.

File: configs/config-mini-tls1_1.h [hunk 13]
```diff
@@ -57,6 +57,7 @@
 #define MBEDTLS_RSA_C
 #define MBEDTLS_SHA1_C
 #define MBEDTLS_SHA256_C
+#define MBEDTLS_SHA224_C
```
[the remark this answers is not in the document]
Comment (mstarzyk-mobica) [reply configs/config-mini-tls1_1.h:59]: :rocket:

File: configs/config-psa-crypto.h [hunk 14]
```diff
@@ -2615,7 +2626,7 @@
 /**
  * \def MBEDTLS_SHA256_C
  *
- * Enable the SHA-224 and SHA-256 cryptographic hash algorithms.
+ * Enable the SHA-256 cryptographic hash algorithms.
```
[the remark this answers is not in the document]
Comment (mstarzyk-mobica) [reply configs/config-psa-crypto.h]: file was removed

File: configs/config-ccm-psk-tls1_2.h (hunk 9 above)
[the remark this answers is not in the document]
Comment (gilles-peskine-arm) [reply configs/config-ccm-psk-tls1_2.h:51]: Yes, please add it to all the configs that now have an explicit `MBEDTLS_SHA224_C`.

**Review by ronald-cron-arm — CHANGES_REQUESTED** [review]
Thanks for the changes, getting there. It seems there are still code guarded by MBEDTLS_SHA512_C in oid.c and ssl_tls.c.
I ran again rgrep -n2 MBEDTLS_SHA512_C and searching for 384 I hit again some suspicious cases in oid.c and ssl_tls.c.
Furthermore the following are suspicious as well:
```
tests/suites/test_suite_shax.function-206-/* END_CASE */                        
tests/suites/test_suite_shax.function-207-                                      
tests/suites/test_suite_shax.function:208:/* BEGIN_CASE depends_on:MBEDTLS_SHA512_C */
tests/suites/test_suite_shax.function-209-void sha384( data_t * src_str, data_t * hash )
tests/suites/test_suite_shax.function-210-{
--                                                                              
library/ssl_misc.h-155-#if defined(MBEDTLS_SSL_SOME_MODES_USE_MAC)              
library/ssl_misc.h-156-/* Ciphersuites using HMAC */                            
library/ssl_misc.h:157:#if defined(MBEDTLS_SHA512_C)                            
library/ssl_misc.h-158-#define MBEDTLS_SSL_MAC_ADD                 48  /* SHA-384 used for HMAC */
library/ssl_misc.h-159-#elif defined(MBEDTLS_SHA256_C)                          
--                                                                              
library/ssl_misc.h-551-#endif                                                   
library/ssl_misc.h-552-#endif                                                   
library/ssl_misc.h:553:#if defined(MBEDTLS_SHA512_C)                            
library/ssl_misc.h-554-#if defined(MBEDTLS_USE_PSA_CRYPTO)                      
library/ssl_misc.h-555-    psa_hash_operation_t fin_sha384_psa;
```

File: tests/suites/test_suite_rsa.data (hunk 6 above)
[the remark this answers is not in the document]
Comment (gilles-peskine-arm) [reply tests/suites/test_suite_rsa.data]: I ran `tests/scripts/analyze_outcome.py` on the outcome file of https://jenkins-mbedtls.oss.arm.com/job/mbed-tls-pr-merge/job/PR-4304-merge/31/ which ran on 5d5a59895f19827865e54c2873f61c8416ed381a merged with `development` at eb3e463380334687d431b894a0dfe7bfe9d42fd2, and the outcome file of https://jenkins-mbedtls.oss.arm.com/job/mbed-tls-nightly-tests/1635/ which ran on `development` at eb3e463380334687d431b894a0dfe7bfe9d42fd2. The two logs are identical. So this pull request does not introduce any unmet dependencies.

**Review by gilles-peskine-arm — CHANGES_REQUESTED** [review]
Looks good to me except for a few points:
* There are a few English mistakes in `config.h` (a very visible part of the library's documentation).
* The [dependency issues that Ronald found](https://github.com/ARMmbed/mbedtls/pull/4304#pullrequestreview-656975612) aren't blockers (we can resolve them when we start supporting SHA256 without SHA224 and so on), but we might as well fix the issues that we're aware of.
* Commit messages and commit structure issues as explained below.
* [Changelog and migration guide](https://github.com/ARMmbed/mbedtls/pull/4304/files#r631370131)
* More [SHA384 guards in TLS](https://github.com/ARMmbed/mbedtls/pull/4304/files#r631390326)

Some notes on commit structure:
* 6326a8dd37990beff4b488aff1ea7f9138edcf91 (“Fix SHA definitions and their dependencies in library and test suites.”) was very hard to follow. It should have been broken into multiple commits:
    * One to reorder entries in `config.h` (and re-run `check-generated-files.sh -u`). (`git show --color-moved` would only show moved things.)
    * One or more to modify entries in `config.h`.
    * One for the changes to `configs/*`.
    * One or more for the miscellaneous dependency changes in `include/mbedtls/*.h` (excluding `config.h`) and `library`.
    * One for the test dependency updates.
* The merge commit 667de6c1b3243648ae638c4c67b7d441c347c563 should have an explanation of the conflict resolution in the commit message. See https://github.com/gilles-peskine-arm/mbedtls/tree/pr_4304-merge-development-20210510 for an example of what I consider an adequate explanation.

I would appreciate breaking 6326a8dd37990beff4b488aff1ea7f9138edcf91 down into manageable pieces, but because I know it's somewhat inconvenient to rebase beyond a merge, I won't make it a blocker. Please do edit the description of the merge commit to document the conflict resolution. Do not change the content that goes into the merge.

File: library/ssl_ciphersuites.c [hunk 15]
```diff
@@ -395,7 +395,7 @@ static const mbedtls_ssl_ciphersuite_t ciphersuite_definitions[] =
       0 },
 #endif /* MBEDTLS_GCM_C */
 #endif /* MBEDTLS_SHA256_C */
-#if defined(MBEDTLS_SHA512_C)
+#if defined(MBEDTLS_SHA384_C)
```
Comment (gilles-peskine-arm) [inline library/ssl_ciphersuites.c:398]: This fixes https://github.com/ARMmbed/mbedtls/issues/4499.

This pull request needs:
* Changelog entries for the fixes of https://github.com/ARMmbed/mbedtls/issues/4472 and https://github.com/ARMmbed/mbedtls/issues/4499.
* Changelog entries for the changes in configuration options.
* A [migration guide entry](https://github.com/ARMmbed/mbedtls/blob/development/docs/3.0-migration-guide.d/00README).

File: library/ssl_tls.c [hunk 16]
```diff
@@ -6820,15 +6828,19 @@ unsigned char mbedtls_ssl_hash_from_md_alg( int md )
...
 #endif
-#if defined(MBEDTLS_SHA512_C)
+#if defined(MBEDTLS_SHA384_C)
         case MBEDTLS_MD_SHA384:
             return( MBEDTLS_SSL_HASH_SHA384 );
+#endif
+#if defined(MBEDTLS_SHA512_C)
         case MBEDTLS_MD_SHA512:
             return( MBEDTLS_SSL_HASH_SHA512 );
 #endif
```
Comment (gilles-peskine-arm) [inline library/ssl_tls.c:6846]: All the other conditional guards on `MBEDTLS_SHA512_C` are about SHA-384 and should change to `MBEDTLS_SHA384_C`.

File: include/mbedtls/config.h [hunk 17]
```diff
@@ -3106,9 +3106,9 @@
  * \def MBEDTLS_SHA224_C
  *
  * Enable the SHA-224 cryptographic hash algorithm.
- * Future version will have this option disabled by default.
+ * A future version of the library will have this option disabled by default.
```
Comment (gilles-peskine-arm) [inline include/mbedtls/config.h:3109]: I didn't pick up on this sooner, but we don't actually intend to disable SHA-224 by default. The default configuration generally includes all cryptographic mechanisms. It's only in the sample configs in `configs/*` that we intend to disable SHA-224 (because they're sample configs that require SHA-256 only).

**Review by gilles-peskine-arm — CHANGES_REQUESTED** [review]
Looks good to me except for one sentence.

File: include/mbedtls/config.h (hunk 17 above)
Comment (mstarzyk-mobica) [reply include/mbedtls/config.h:3109]: Oh. My bad, will fix tomorrow :)

**Review by gilles-peskine-arm — APPROVED** [review]

**Review by ronald-cron-arm — CHANGES_REQUESTED** [review]
I just spotted three minor things.

File: include/mbedtls/sha512.h [hunk 18]
```diff
@@ -101,8 +101,8 @@ void mbedtls_sha512_clone( mbedtls_sha512_context *dst,
  * \param is384    Determines which function to use. This must be
  *                 either \c 0 for SHA-512, or \c 1 for SHA-384.
  *
- * \note           When \c MBEDTLS_SHA512_NO_SHA384 is defined, \p is384 must
- *                 be \c 0, or the function will return
+ * \note           When \c MBEDTLS_SHA384_C is defined,
```
[the remark this answers is not in the document]
Comment (mstarzyk-mobica) [reply include/mbedtls/sha512.h]: Ah, should be 'not defined'

**Review by ronald-cron-arm — APPROVED** [review]
All good on my side now. Thanks.

**Review by gilles-peskine-arm — APPROVED** [review]

File: docs/3.0-migration-guide.d/modify_SHA384_option_behaviour.md [hunk 19]
```diff
@@ -0,0 +1,9 @@
+Replaced MBEDTLS_SHA512_NO_SHA384 with MBEDTLS_SHA384_C
+------------------------------------------------------
+
+This does not affect users who use the default `config.h`.
+MBEDTLS_SHA512_NO_SHA384 was disabled by default, now MBEDTLS_SHA384_C is
+enabled by default.
+
```
Comment (gilles-peskine-arm) [inline docs/3.0-migration-guide.d/modify_SHA384_option_behaviour.md:7]: I just noticed something important is missing: users who had their own configuration file with `MBEDTLS_SHA512_C` enabled and not `MBEDTLS_SHA512_NO_SHA384`, and who need SHA-384, now need to explicitly enable `MBEDTLS_SHA384_C`.

**Review by gilles-peskine-arm — CHANGES_REQUESTED** [review]
I just spotted an omission in the migration guide.

**Review by gilles-peskine-arm — APPROVED** [review]
Please merge `development` into your branch to resolve the conflict with the removal of generated files.

**Review by gilles-peskine-arm — APPROVED** [review]

**Review by ronald-cron-arm — APPROVED** [review]
I did the merge on my side and ended up with the same tree thus still good to me.
